### DISCUSSION TITLE
Cascading event hooks

### DISCUSSION THREAD

**Discussion by akiarostami** [pocketbase/pocketbase#4383] [Q&A] [answered]
Hello all,

I have two "After Update" event hooks for two collections.

In the AfterUpdate1 (only set to monitor Collection 1), when a field changes in Collection1, I update the status of related record in Collection2 to "changed".

In the AfterUpdate12 (only set to monitor Collection 2), when status becomes "changed", I set the "changed_at" field to the current time (and do some other stuff).

AfterUpdate2 triggers when I make a change in Collection2 via Admin or API, but when I change the Collection2's status inside AfterUpdate1 event, AfterUpdate2 doesn't get triggered.

Is there a way to have cascading event hooks for a scenario like this? If not, is there a way to call an even hook inside another event hook?

Thank you!

**Comment by ganigeorgiev** [comment]
There are too many factors in play here. Please provide a code sample with your event hooks.

**Reply by akiarostami** [reply]
Sure. Here's the code:

```
onRecordAfterCreateRequest((e) => {
    $app.dao().db()
        .newQuery('UPDATE articles SET status='edited' WHERE id = {:id}')
        .bind({ "id": e.record.getString("article") })
        .execute();
}, "comments");

onRecordAfterUpdateRequest((e) => {
    let record = e.record;
    let status = record.getString('status');
    if (status === 'edited') {
      record.set('last_edit', new Date());
      $app.dao().saveRecord(record);
    }
}, "articles");
```

If I manually change the status of an article, the `last_edit` field gets updated. 

When I create a new comment, the status of the article changes to `edited`, but the `last_edit` field won't change (onRecordAfterUpdateRequest won't get triggered). 

For the `onRecordAfterCreateRequest`, I also tried setting the value using record update / save methods, and I got the same result; status was updated but `AfterUpdate` trigger wasn't fired.

**Reply by ganigeorgiev** [reply]
Hooks are not fired when you execute raw queries.
Update/Create/Delete hooks for the records are triggered when you operate with the `Record` model.

**Reply by ganigeorgiev** [reply]
Edit: On second read, the `onRecordAfterUpdateRequest` will note get triggered even if you update each record with `saveRecord()` in your `onRecordAfterCreateRequest` hook because the `*Request` hooks are fired **only** when performing API requests.

To listen to general DB model changes you can attach to the DB model hooks - https://pocketbase.io/docs/js-event-hooks/#db-hooks.

**Reply by ganigeorgiev** [reply, marked as answer]
Here is a slightly modified version of your example:

```js
onModelAfterCreate((e) => {
    const record = e.dao.findRecordById("articles", e.model.get("article"))
    if (record.get("status") != "edited") {
        record.set("status", "edited")
        e.dao.saveRecord(record);
    }
}, "comments")

// note: eventually you can attach to the onModelBeforeUpdate hook and skip the extra saveRecord
onModelAfterUpdate((e) => {
    if (e.model.get("status") == "edited") {
        e.model.set("last_edit", new Date());  // technically raw Date value are not supported but I guess it would be casted to the correct type (types.DateTime)
        // save without hooks to prevent infinite recursion
        e.dao.withoutHooks().saveRecord(e.model);
    }
}, "articles")
```
_edit: updated to use the event `e.dao` (in this case it doesn't matter because it is an after hook)_
_edit2: updated `onModelAfterCreate` to use `findRecordById` since you are targetting a single record by their id._

**Reply by akiarostami** [reply]
Thanks for the prompt and clear response! This solved my problem.
